File: 572. Subtree of Another Tree/572. Subtree of Another Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

  public boolean isSubtree(TreeNode root, TreeNode subRoot) {
    return hasSubtree(root, subRoot);
  }

  private boolean isSameTree(TreeNode p, TreeNode q) {
    if (p == null && q == null) {
      return true;
    }

    if (
      (p != null && q == null) || (p == null && q != null) || p.val != q.val
    ) {
      return false;
    }

    return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
  }

  private boolean hasSubtree(TreeNode p, TreeNode subRoot) {
    if (p == null) {
      return false;
    }

    if (isSameTree(p, subRoot)) {
      return true;
    }

    return hasSubtree(p.left, subRoot) || hasSubtree(p.right, subRoot);
  }
}

File: 572. Subtree of Another Tree/572. Subtree of Another Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def is_same_tree(p, q):
            if not p and not q:
                return True

            if not p and q or p and not q or p.val != q.val:
                return False

            return is_same_tree(p.left, q.left) and is_same_tree(p.right, q.right)

        def has_subtree(p):
            if not p:
                return False

            if is_same_tree(p, subRoot):
                return True

            return has_subtree(p.left) or has_subtree(p.right)

        return has_subtree(root)
